pkg/webhook/webhook.go

```diff
@@ -6,95 +6,48 @@ import (
 	"io"
 	"net/http"
 
-	"github.com/radiofrance/gitlab-ci-pipelines-exporter/pkg/collectors"
-
-	"github.com/prometheus/client_golang/prometheus"
-	"github.com/prometheus/client_golang/prometheus/promhttp"
+	gcpehttp "github.com/radiofrance/gitlab-ci-pipelines-exporter/pkg/http"
+	"github.com/radiofrance/gitlab-ci-pipelines-exporter/pkg/metrics"
 	"github.com/urfave/negroni"
 	"go.uber.org/zap"
 )
 
 type (
-	// Collectors groups all Prometheus collectors used to exporter Gitlab CI metrics.
-	Collectors struct {
-		IDCollector                       *prometheus.GaugeVec
-		DurationSecondsCollector          *prometheus.GaugeVec
-		QueuedDurationSecondsCollector    *prometheus.GaugeVec
-		RunCountCollector                 *prometheus.CounterVec
-		StatusCollector                   *prometheus.GaugeVec
-		TimestampCollector                *prometheus.GaugeVec
-		JobIDCollector                    *prometheus.GaugeVec
-		JobDurationSecondsCollector       *prometheus.GaugeVec
-		JobQueuedDurationSecondsCollector *prometheus.GaugeVec
-		JobRunCountCollector              *prometheus.CounterVec
-		JobStatusCollector                *prometheus.GaugeVec
-		JobTimestampCollector             *prometheus.GaugeVec
-	}
-
 	Webhook struct {
-		http.Handler
-		Collectors
-
+		metrics.Collectors
+		mux *http.ServeMux
 		log *zap.Logger
 	}
 )
 
 // NewWebhook creates a new instance of the Gitlab event webhook handler.
-func NewWebhook(pattern, secretToken string, opts ...Option) *Webhook {
+func NewWebhook(secretToken string, opts ...Option) *Webhook {
 	webhook := &Webhook{
-		Handler: http.NewServeMux(),
-		Collectors: Collectors{
-			IDCollector:                       collectors.NewCollectorID(),
-			DurationSecondsCollector:          collectors.NewCollectorDurationSeconds(),
-			QueuedDurationSecondsCollector:    collectors.NewCollectorQueuedDurationSeconds(),
-			RunCountCollector:                 collectors.NewCollectorRunCount(),
-			StatusCollector:                   collectors.NewCollectorStatus(),
-			TimestampCollector:                collectors.NewCollectorTimestamp(),
-			JobIDCollector:                    collectors.NewCollectorJobID(),
-			JobDurationSecondsCollector:       collectors.NewCollectorJobDurationSeconds(),
-			JobQueuedDurationSecondsCollector: collectors.NewCollectorJobQueuedDurationSeconds(),
-			JobRunCountCollector:              collectors.NewCollectorJobRunCount(),
-			JobStatusCollector:                collectors.NewCollectorJobStatus(),
-			JobTimestampCollector:             collectors.NewCollectorJobTimestamp(),
-		},
-
-		log: zap.Must(zap.NewProduction()),
+		Collectors: metrics.AllCollectors(),
```

Comment:
I suggest moving the collectors as parameters; I find it a bit odd to generate all the collectors in the webhook and then inject them into the metrics handler.

Reply:
I agree it not the handler's responsibility